### PR TITLE
Add support conditions for duplicated probes

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Keep track of capturing probeIds based on condition evaluations before adding snapshots to the sink

# Motivation
Support for conditions for duplicated probes

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java [hunk 1]
```diff
@@ -164,21 +168,29 @@ public void commit() {
...
+                version,
+                timestamp,
+                duration,
+                stack,
+                captures,
+                new ProbeDetails(additionalProbe.id, probe.location, probe.script, probe.tags),
+                language,
+                thread,
+                traceId,
+                spanId));
```
Comment (cimi) [inline dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java:190]: Nit: we could add a `.copy(String newId) -> Snapshot` method to the Snapshot class if we're doing this in more than one place.

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java [hunk 2]
```diff
@@ -188,19 +200,37 @@ public boolean isCapturing() {
...
+    if (isCapturing) {
+      capturingProbeIds.add(probe.id);
+    }
+    List<ProbeDetails> additionalProbes = probe.additionalProbes;
+    if (!additionalProbes.isEmpty()) {
+      for (ProbeDetails additionalProbe : additionalProbes) {
+        if (executeScript(additionalProbe.getScript(), capture, additionalProbe.getId())) {
+          isCapturing = true; // force capturing the current Snapshot
+          capturingProbeIds.add(additionalProbe.getId());
+        }
```
Comment (cimi) [inline dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java:214]: I don't understand this part - why do we need `isCapturing` if we keep `capturingProbeIds`? Isn't it equivalent to checking if the collection has length zero?

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/CapturedSnapshotTest.java [hunk 3]
```diff
@@ -809,10 +805,58 @@ public void simpleConditionTest() throws IOException, URISyntaxException {
...
+        listener.snapshots.get(0).getCaptures().getEntry(), "arg", String.class.getName(), "5");
+  }
+
+  @Test
+  public void mergedProbesWithAllConditionsFalseTest() throws IOException, URISyntaxException {
+    doMergedProbeConditions(
+        new ProbeCondition(DSL.when(DSL.FALSE), "false"),
+        new ProbeCondition(DSL.when(DSL.FALSE), "false"),
+        0);
+  }
```
Comment (cimi) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/CapturedSnapshotTest.java:818]: Nit: we could add a test where both conditions are `true` - we do test that two snapshots are emitted when we have a true condition and a probe without, so I guess that covers most of it.

**Review by cimi — APPROVED** [review]
Thanks! 

Left a couple of comments, nothing blocking.

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java (hunk 1 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java]: done

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java (hunk 2 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java]: I was reluctant to remove this field as it is used by instrumentation, but you're right we can just rely on the set.

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/CapturedSnapshotTest.java (hunk 3 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/CapturedSnapshotTest.java:826]: done

**Review by cimi — APPROVED** [review]
:+1: Thanks!

**Pull request merged**